codegenerator/cli/templates/static/codegen/src/eventFetching/SourceManager.res

```diff
@@ -240,3 +252,103 @@ let waitForNewBlock = async (sourceManager: t, ~currentBlockHeight) => {
+  }
+
+  let responseRef = ref(None)
+
+  while responseRef.contents->Option.isNone {
+    let source = sourceManager.activeSource
+
+    let logger = Logging.createChild(
+      ~params={
+        "chainId": source.chain->ChainMap.Chain.toChainId,
+        "logType": "Block Range Query",
+        "partitionId": query.partitionId,
+        "source": source.name,
+        "fromBlock": query.fromBlock,
+        "toBlock": toBlock,
+        "addresses": addresses,
+      },
+    )
+
+    try {
+      let response = await source.getItemsOrThrow(
+        ~fromBlock=query.fromBlock,
+        ~toBlock,
+        ~contractAddressMapping=query.contractAddressMapping,
+        ~partitionId=query.partitionId,
+        ~currentBlockHeight,
+        ~selection=query.selection,
+        ~logger,
+      )
+      logger->Logging.childTrace({
+        "msg": "Fetched block range from server",
+        "latestFetchedBlockNumber": response.latestFetchedBlockNumber,
+        "numEvents": response.parsedQueueItems->Array.length,
+        "stats": response.stats,
+      })
+      responseRef := Some(response)
+    } catch {
+    | Source.GetItemsError(error) => {
+        // TODO: When we start handling fetch failures,
+        // we shouldn't delete the source from the set
+        // but keep it for retries. Still need to delete
+        // for cases like UnsupportedSelection which are
+        // not retryable
+        let notAlreadyDeleted = sourceManager.sources->Utils.Set.delete(source)
+
+        // In case there are multiple partitions
+        // failing at the same time. Log only once
+        if notAlreadyDeleted {
+          switch error {
+          | UnsupportedSelection({message}) => logger->Logging.childError(message)
+          | FailedGettingFieldSelection({message, blockNumber, logIndex})
+          | FailedParsingItems({message, blockNumber, logIndex}) =>
+            logger->Logging.childError({
+              "msg": message,
+              "blockNumber": blockNumber,
+              "logIndex": logIndex,
+            })
+          }
+        }
+
+        switch sourceManager->getNextActiveSource {
+        | None => {
+            logger->Logging.childError(
+              "The indexer doesn't have data-sources which can continue fetching. Please, check the error logs or reach out to the Envio team.",
+            )
+            // Hang until the process is terminated
+            await Promise.make((_, _) => ())
+          }
+        | Some(nextSource) => {
+            logger->Logging.childInfo({
+              "msg": "Switching to another data-source",
+              "source": nextSource.name,
+            })
+            sourceManager.activeSource = nextSource
+            // Will loop because of while
+          }
+        }
+      }
+    // TODO: Handle more error cases and hang/retry instead of throwing
+    | exn => exn->ErrorHandling.mkLogAndRaise(~logger, ~msg="Failed to fetch block Range")
```

Comment:
HyperSync will fail as before

Reply:
Should I catch it and `dispatchAction(ErrorExit(errHandler))` or simply throwing is fine?